File: app/src/main/java/pl/pzienowicz/zditmszczecinlive/activity/MainActivity.java
package pl.pzienowicz.zditmszczecinlive.activity;

import android.Manifest;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.preference.PreferenceManager;
import android.support.design.widget.Snackbar;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.LinearLayout;

import com.getbase.floatingactionbutton.FloatingActionButton;
import com.getbase.floatingactionbutton.FloatingActionsMenu;
import com.onesignal.OneSignal;

import java.util.Timer;
import java.util.TimerTask;

import pl.pzienowicz.zditmszczecinlive.Config;
import pl.pzienowicz.zditmszczecinlive.Functions;
import pl.pzienowicz.zditmszczecinlive.R;
import pl.pzienowicz.zditmszczecinlive.dialog.LineDialog;
import pl.pzienowicz.zditmszczecinlive.dialog.SettingsDialog;

public class MainActivity extends AppCompatActivity implements LocationListener {

    private WebView myWebView = null;
    private SharedPreferences sharedPreferences = null;
    private Context context = null;
    private BroadcastReceiver bcr = null;
    private LocationManager locationManager = null;
    private Location currentLocation = null;
    private Timer myTimer = null;
    private boolean zoomMap = false;
    private String currentUrl = Config.URL;

    private static final int MY_PERMISSIONS_REQUEST_LOCATION = 1443;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        OneSignal.startInit(this).init();

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        context = this;

        if (getSupportActionBar() != null) {
            getSupportActionBar().hide();
        }

        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
        sharedPreferences.edit().putInt(Config.PREFERENCE_SELECTED_LINE, 0).apply();

        final FloatingActionsMenu floatingActionsMenu = (FloatingActionsMenu) findViewById(R.id.multiple_actions);

        FloatingActionButton setFavouriteBtn = (FloatingActionButton) findViewById(R.id.set_favourite);
        setFavouriteBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sharedPreferences.edit().putString(Config.PREFERENCE_FAVOURITE_MAP, currentUrl).apply();
                Snackbar.make(findViewById(R.id.swiperefresh), R.string.set_favourite, Snackbar.LENGTH_LONG).show();
                floatingActionsMenu.collapse();
            }
        });

        FloatingActionButton actionButton = (FloatingActionButton) findViewById(R.id.show_lines);
        actionButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                LineDialog dialog = new LineDialog(context);
                dialog.getWindow().setLayout(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
                dialog.show();
                floatingActionsMenu.collapse();
            }
        });

        FloatingActionButton settingsButton = (FloatingActionButton) findViewById(R.id.settings);
        settingsButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                SettingsDialog dialog = new SettingsDialog(context);
                dialog.getWindow().setLayout(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
                dialog.show();
                floatingActionsMenu.collapse();
            }
        });

        myWebView = (WebView) findViewById(R.id.webView);
        myWebView.getSettings().setJavaScriptEnabled(true);
        myWebView.setWebViewClient(new WebViewClient() {
            public void onPageFinished(WebView view, String url) {
            }
        });

        IntentFilter filter = new IntentFilter();
        filter.addAction(Config.INTENT_LOAD_NEW_URL);
        filter.addAction(Config.INTENT_REFRESH_SETTINGS);
        filter.addAction(Config.INTENT_NO_INTERNET_CONNETION);

        bcr = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                switch (intent.getAction()) {
                    case Config.INTENT_LOAD_NEW_URL:
                        int lineId = intent.getIntExtra(Config.EXTRA_LINE_ID, 0);
                        if (lineId == 0) {
                            currentUrl = Config.URL;
                        } else {
                            currentUrl = Config.LINE_URL + lineId;
                        }
                        myWebView.loadUrl(currentUrl);
                        break;
                    case Config.INTENT_REFRESH_SETTINGS:
                        refreshSettings();
                        break;
                    case Config.INTENT_NO_INTERNET_CONNETION:
                        showNoInternetSnackbar();
                        break;
                }
            }
        };
        registerReceiver(bcr, filter);
        refreshSettings();

        if (savedInstanceState != null) {
            myWebView.restoreState(savedInstanceState);
        } else {
            loadPage();
        }

        myTimer = new Timer();
        myTimer.scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                if(currentLocation != null) {
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            String newUrl = currentUrl + "?lat=" + currentLocation.getLatitude() + "&lon=" + currentLocation.getLongitude();

                            if(zoomMap) {
                                newUrl += "&zoom=16";
                            }

                            myWebView.loadUrl(newUrl);
                        }
                    });
                }
            }
        }, 0, 30 * 1000);
    }

    private void refreshSettings() {
        locationManager = (LocationManager) getSystemService(LOCATION_SERVICE);

        if (sharedPreferences.getBoolean(Config.PREFERENCE_USE_LOCATION, true)) {
            if (ActivityCompat.checkSelfPermission(MainActivity.this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
                locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, MainActivity.this);
            } else {
                ActivityCompat.requestPermissions(MainActivity.this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, MY_PERMISSIONS_REQUEST_LOCATION);
            }
        } else {
            if(locationManager != null) {
                locationManager.removeUpdates(MainActivity.this);
                locationManager = null;
            }
            currentLocation = null;
        }

        zoomMap = sharedPreferences.getBoolean(Config.PREFERENCE_ZOOM_MAP, true);
    }

    @Override
    public void onDestroy() {
        if(myTimer != null) {
            myTimer.cancel();
        }
        unregisterReceiver(bcr);
        super.onDestroy();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        myWebView.saveState(outState);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (event.getAction() == KeyEvent.ACTION_DOWN) {
            switch (keyCode) {
                case KeyEvent.KEYCODE_BACK:
                    finish();
                    return true;
            }

        }
        return super.onKeyDown(keyCode, event);
    }

    private void loadPage() {
        if (Functions.isNetworkAvailable(this)) {
            String favouriteMap = sharedPreferences.getString(Config.PREFERENCE_FAVOURITE_MAP, Config.URL);

            myWebView.loadUrl(favouriteMap);
            showInitDialog();
        } else {
            showNoInternetSnackbar();
        }
    }

    private void showInitDialog() {
        if (!sharedPreferences.getBoolean(Config.PREFERENCE_SHOW_DIALOG, true)) {
            return;
        }

        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(this);
        alertDialogBuilder.setTitle(R.string.dialog_title);
        alertDialogBuilder
                .setMessage(R.string.info)
                .setCancelable(false)
                .setPositiveButton(R.string.close, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        dialog.cancel();
                    }
                })
                .setNegativeButton(R.string.do_not_show_more, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        sharedPreferences.edit().putBoolean(Config.PREFERENCE_SHOW_DIALOG, false).apply();
                        dialog.cancel();
                    }
                });

        AlertDialog alertDialog = alertDialogBuilder.create();
        alertDialog.show();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        switch (requestCode) {
            case MY_PERMISSIONS_REQUEST_LOCATION: {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                        sharedPreferences.edit().putBoolean(Config.PREFERENCE_ZOOM_MAP, false).apply();
                        sharedPreferences.edit().putBoolean(Config.PREFERENCE_USE_LOCATION, false).apply();
                    }
                    locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, MainActivity.this);
                } else {
                    sharedPreferences.edit().putBoolean(Config.PREFERENCE_ZOOM_MAP, false).apply();
                    sharedPreferences.edit().putBoolean(Config.PREFERENCE_USE_LOCATION, false).apply();
                }
                break;
            }
        }
    }


    private void showNoInternetSnackbar()
    {
        Snackbar
                .make(findViewById(R.id.swiperefresh), R.string.no_internet, Snackbar.LENGTH_INDEFINITE)
                .setAction(R.string.refresh, new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        loadPage();
                    }
                })
                .show();
    }

    @Override
    public void onLocationChanged(Location location) {
        if(location != null) {
            currentLocation = location;
        }
    }

    @Override
    public void onStatusChanged(String s, int i, Bundle bundle) {

    }

    @Override
    public void onProviderEnabled(String s) {

    }

    @Override
    public void onProviderDisabled(String s) {

    }
}

File: app/src/main/java/pl/pzienowicz/zditmszczecinlive/activity/MainActivity.kt
package pl.pzienowicz.zditmszczecinlive.activity

import android.Manifest
import android.annotation.SuppressLint
import android.app.Dialog
import android.content.BroadcastReceiver
import android.content.ActivityNotFoundException
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.location.Location
import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.webkit.GeolocationPermissions
import android.webkit.WebResourceRequest
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.splashscreen.SplashScreen.Companion.installSplashScreen
import androidx.core.app.ActivityCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import androidx.core.view.updatePadding
import android.widget.PopupWindow
import pl.pzienowicz.zditmszczecinlive.BuildConfig
import pl.pzienowicz.zditmszczecinlive.*
import pl.pzienowicz.zditmszczecinlive.billing.GooglePlayBillingClient
import pl.pzienowicz.zditmszczecinlive.data.FavouritesRepository
import pl.pzienowicz.zditmszczecinlive.data.PremiumLimits
import pl.pzienowicz.zditmszczecinlive.databinding.ActivityMainBinding
import pl.pzienowicz.zditmszczecinlive.dialog.*
import pl.pzienowicz.zditmszczecinlive.model.FavouriteLine
import pl.pzienowicz.zditmszczecinlive.timer.MapTimer
import androidx.core.net.toUri

class MainActivity : AppCompatActivity() {

    private var bcr: BroadcastReceiver? = null
    private var currentLocation: Location? = null
    private lateinit var mapTimer: MapTimer
    private var currentUrl = Config.URL
    var mGeoLocationCallback: GeolocationPermissions.Callback? = null
    var mGeoLocationRequestOrigin: String? = null

    private lateinit var binding: ActivityMainBinding
    private lateinit var billingClient: GooglePlayBillingClient
    private val premiumLimits = PremiumLimits()
    private var premiumUnlockedAfterPurchase = false
    private var morePopupWindow: PopupWindow? = null

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreate(savedInstanceState: Bundle?) {
        installSplashScreen()
        super.onCreate(savedInstanceState)
        supportActionBar?.hide()

        @Suppress("DEPRECATION")
        window.statusBarColor = Color.TRANSPARENT
        WindowCompat.setDecorFitsSystemWindows(window, false)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        applyWindowInsets()

        WindowInsetsControllerCompat(window, window.decorView).hide(WindowInsetsCompat.Type.statusBars())

        prefs.selectedLine = 0
        migrateFavouriteMap()
        billingClient = GooglePlayBillingClient(
            activity = this,
            onInitialized = {},
            onPurchased = {
                premiumUnlockedAfterPurchase = true
                showBar(R.string.payment_success)
            }
        )

        binding.setFavourite.setOnClickListener {
            toggleFavouriteLine()
            updateFavouriteIcon()
        }
        updateFavouriteIcon()

        binding.navLines.setOnClickListener {
            openLineDialog()
        }

        binding.navInfo.setOnClickListener {
            openInfoDialog()
        }

        binding.navDashboard.setOnClickListener {
            openScheduleBoardDialog()
        }

        binding.navFavourites.setOnClickListener {
            openFavouritesDialog()
        }
        if (BuildConfig.DEBUG) {
            binding.navFavourites.setOnLongClickListener {
                billingClient.consumePremiumProducts()
                true
            }
        }

        binding.navMore.setOnClickListener { showMoreMenu() }

        binding.webView.settings.javaScriptEnabled = true
        binding.webView.settings.domStorageEnabled = true
        binding.webView.settings.setGeolocationEnabled(true)
        binding.webView.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(
                view: WebView?,
                request: WebResourceRequest?
            ): Boolean = openLinkExternallyIfNeeded(request?.url?.toString())

            @Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")
            override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean =
                openLinkExternallyIfNeeded(url)

            override fun onPageFinished(view: WebView?, url: String?) {
                applyMapThemePreference()
            }
        }
        binding.webView.webChromeClient = object : WebChromeClient() {
            override fun onGeolocationPermissionsShowPrompt(
                origin: String, callback: GeolocationPermissions.Callback
            ) {
                if (
                    ActivityCompat.checkSelfPermission(
                        applicationContext,
                        Manifest.permission.ACCESS_FINE_LOCATION
                    ) != PackageManager.PERMISSION_GRANTED &&
                    ActivityCompat.checkSelfPermission(
                        applicationContext,
                        Manifest.permission.ACCESS_COARSE_LOCATION
                    ) != PackageManager.PERMISSION_GRANTED
                ) {
                    mGeoLocationCallback = callback
                    mGeoLocationRequestOrigin = origin
                    ActivityCompat.requestPermissions(
                        this@MainActivity,
                        arrayOf(
                            Manifest.permission.ACCESS_FINE_LOCATION,
                            Manifest.permission.ACCESS_COARSE_LOCATION
                        ),
                        MY_PERMISSIONS_REQUEST_LOCATION
                    )
                } else {
                    callback.invoke(origin, true, false)
                }
            }
        }
        setupBackNavigation()

        bcr = registerReceiver(listOf(
            Config.INTENT_LOAD_NEW_URL,
            Config.INTENT_REFRESH_SETTINGS,
            Config.INTENT_NO_INTERNET_CONNECTION
        )) { intent ->
            when (intent?.action) {
                Config.INTENT_LOAD_NEW_URL -> {
                    val lineId = intent.getStringExtra(Config.EXTRA_LINE_ID)
                    currentUrl = if (lineId.isNullOrEmpty()) {
                        Config.URL
                    } else {
                        Config.LINE_URL + lineId
                    }
                    Log.d(Config.LOG_TAG, currentUrl)
                    binding.webView.loadUrl(currentUrl)
                    updateFavouriteIcon()
                }
                Config.INTENT_REFRESH_SETTINGS -> applyMapThemePreference()
                Config.INTENT_NO_INTERNET_CONNECTION -> showNoInternetSnackbar()
            }
        }

        if (savedInstanceState != null) {
            binding.webView.restoreState(savedInstanceState)
        } else {
            loadPage()
        }

        mapTimer = MapTimer {
            runOnUiThread {
                currentLocation?.let {
                    val url = currentUrl
                        .plus("?lat=" + it.latitude)
                        .plus("&lon=" + it.longitude)
                    Log.d(Config.LOG_TAG, url)
                    binding.webView.loadUrl(url)
                }
            }
        }
        mapTimer.start()

        openWidgetsDialogIfRequested(intent)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        openWidgetsDialogIfRequested(intent)
    }

    private fun showDialog(dialog: Dialog) {
        dialog.setFullWidth()
        dialog.show()
    }

    private fun applyWindowInsets() {
        val navigationPaddingLeft = binding.bottomNavigation.paddingLeft
        val navigationPaddingTop = binding.bottomNavigation.paddingTop
        val navigationPaddingRight = binding.bottomNavigation.paddingRight
        val navigationPaddingBottom = binding.bottomNavigation.paddingBottom
        val navigationLayoutParams = binding.bottomNavigation.layoutParams
        val navigationWidth = navigationLayoutParams.width
        val favouriteLayoutParams = binding.setFavourite.layoutParams as ViewGroup.MarginLayoutParams
        val favouriteMarginRight = favouriteLayoutParams.rightMargin
        val favouriteMarginBottom = favouriteLayoutParams.bottomMargin

        ViewCompat.setOnApplyWindowInsetsListener(binding.root) { _, windowInsets ->
            val navigationBars = windowInsets.getInsets(WindowInsetsCompat.Type.navigationBars())
            if (isLandscape && navigationWidth > 0) {
                navigationLayoutParams.width = navigationWidth + navigationBars.left
                binding.bottomNavigation.layoutParams = navigationLayoutParams
                binding.bottomNavigation.updatePadding(
                    left = navigationPaddingLeft + navigationBars.left,
                    top = navigationPaddingTop + navigationBars.top,
                    right = navigationPaddingRight,
                    bottom = navigationPaddingBottom + navigationBars.bottom
                )
            } else {
                binding.bottomNavigation.updatePadding(
                    left = navigationPaddingLeft + navigationBars.left,
                    top = navigationPaddingTop,
                    right = navigationPaddingRight + navigationBars.right,
                    bottom = navigationPaddingBottom + navigationBars.bottom
                )
            }
            favouriteLayoutParams.rightMargin = favouriteMarginRight + navigationBars.right
            favouriteLayoutParams.bottomMargin = favouriteMarginBottom + navigationBars.bottom
            binding.setFavourite.layoutParams = favouriteLayoutParams

            windowInsets
        }
        ViewCompat.requestApplyInsets(binding.root)
    }

    public override fun onDestroy() {
        mapTimer.stop()
        unregisterReceiver(bcr)
        super.onDestroy()
    }

    private fun setupBackNavigation() {
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                if (binding.webView.canGoBack()) {
                    binding.webView.goBack()
                    return
                }

                if (isSelectedMapUrl(binding.webView.url)) {
                    finish()
                    return
                }

                binding.webView.loadUrl(currentUrl)
            }
        })
    }

    private fun isSelectedMapUrl(url: String?): Boolean =
        normalizeUrl(url) == normalizeUrl(currentUrl)

    private fun openLinkExternallyIfNeeded(url: String?): Boolean {
        if (!prefs.openLinksInExternalBrowser || isSelectedMapUrl(url) || url.isNullOrBlank()) {
            return false
        }

        return try {
            startActivity(Intent(Intent.ACTION_VIEW, url.toUri()))
            true
        } catch (_: ActivityNotFoundException) {
            false
        }
    }

    private fun normalizeUrl(url: String?): String? =
        url
            ?.substringBefore('#')
            ?.substringBefore('?')
            ?.trimEnd('/')

    override fun onSaveInstanceState(outState: Bundle) {
        binding.webView.saveState(outState)
        return super.onSaveInstanceState(outState)
    }

    private fun loadPage() {
        if (isNetworkAvailable) {
            currentUrl = FavouritesRepository(this).getFavouriteLines().firstOrNull()?.url
                ?: Config.URL
            binding.webView.loadUrl(currentUrl)
            updateFavouriteIcon()
            showInitDialog()
        } else {
            showNoInternetSnackbar()
        }
    }

    private fun migrateFavouriteMap() {
        if (prefs.favouriteMapMigrated) {
            return
        }

        prefs.favouriteMapToMigrate()?.let { favouriteMap ->
            val oldFavouriteMap = normalizeUrl(favouriteMap) ?: Config.URL
            val repository = FavouritesRepository(this)
            if (!repository.isFavouriteLine(oldFavouriteMap)) {
                repository.addFavouriteLine(
                    FavouriteLine(
                        title = oldFavouriteMap.toFavouriteLineTitle(),
                        url = oldFavouriteMap
                    )
                )
            }
        }
        prefs.clearFavouriteMap()
        prefs.favouriteMapMigrated = true
    }

    private fun applyMapThemePreference() {
        binding.webView.applyThemePreference(prefs.darkMode)
    }

    private fun updateFavouriteIcon() {
        val icon = if (FavouritesRepository(this).isFavouriteLine(normalizeUrl(currentUrl) ?: currentUrl)) {
            R.drawable.ic_favorite_white_48dp
        } else {
            R.drawable.ic_favorite_border_white_48dp
        }
        binding.setFavourite.setImageResource(icon)
    }

    private fun toggleFavouriteLine() {
        val line = currentFavouriteLine()
        val repository = FavouritesRepository(this)
        if (repository.isFavouriteLine(line.url)) {
            repository.removeFavouriteLine(line.url)
            showBar(R.string.remove_favourite_line_success)
        } else {
            if (!premiumLimits.canAddFavouriteLine(repository.getFavouriteLines().size, isPremiumUnlocked())) {
                showDialog(
                    PremiumDialog(this) {
                        premiumUnlockedAfterPurchase = true
                    }
                )
                return
            }
            repository.addFavouriteLine(line)
            showBar(R.string.add_favourite_line_success)
        }
    }

    private fun isPremiumUnlocked(): Boolean =
        premiumUnlockedAfterPurchase || billingClient.isPremiumUnlocked()

    private fun currentFavouriteLine(): FavouriteLine {
        val url = normalizeUrl(currentUrl) ?: Config.URL
        return FavouriteLine(
            title = url.toFavouriteLineTitle(),
            url = url
        )
    }

    private fun String.toFavouriteLineTitle(): String {
        if (this == Config.URL) {
            return getString(R.string.default_map)
        }

        if (startsWith(Config.LINE_URL)) {
            val lineNumber = removePrefix(Config.LINE_URL)
                .trimEnd('/')
                .substringAfterLast('/')
            if (lineNumber.isNotBlank()) {
                return getString(R.string.favourite_line_title, lineNumber)
            }
        }

        return getString(R.string.map)
    }

    private fun showInitDialog() {
        if (!prefs.showInitDialog) {
            return
        }

        val alertDialogBuilder = AlertDialog.Builder(this)
        alertDialogBuilder.setTitle(R.string.dialog_title)
        alertDialogBuilder
                .setMessage(R.string.info)
                .setCancelable(false)
                .setPositiveButton(R.string.close) { dialog, id -> dialog.cancel() }
                .setNegativeButton(R.string.do_not_show_more) { dialog, id ->
                    prefs.showInitDialog = false
                    dialog.cancel()
                }

        val alertDialog = alertDialogBuilder.create()
        alertDialog.show()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            MY_PERMISSIONS_REQUEST_LOCATION -> {
                // If request is cancelled, the result arrays are empty.
                val fineLocationIndex = permissions.indexOf(Manifest.permission.ACCESS_FINE_LOCATION)
                val coarseLocationIndex = permissions.indexOf(Manifest.permission.ACCESS_COARSE_LOCATION)
                val isLocationGranted = fineLocationIndex >= 0 &&
                    grantResults.getOrNull(fineLocationIndex) == PackageManager.PERMISSION_GRANTED ||
                    coarseLocationIndex >= 0 &&
                    grantResults.getOrNull(coarseLocationIndex) == PackageManager.PERMISSION_GRANTED

                mGeoLocationCallback?.invoke(mGeoLocationRequestOrigin, isLocationGranted, false)
            }
        }
    }

    private fun showNoInternetSnackbar() {
        showBar(R.string.no_internet, R.string.refresh) { loadPage() }
    }

    private fun openInfoDialog() {
        showDialog(InfoDialog(this))
    }

    private fun openScheduleBoardDialog() {
        val dialog = BusStopDialog(this, { busStop ->
            showDialog(ScheduleBoardDialog(this, busStop))
        }, null)
        showDialog(dialog)
    }

    private fun openLineDialog() {
        showDialog(LineDialog(this))
    }

    private fun openFavouritesDialog() {
        showDialog(
            FavouritesDialog(
                activity = this,
                onLineSelected = { line ->
                    currentUrl = line.url
                    binding.webView.loadUrl(line.url)
                    updateFavouriteIcon()
                }
            )
        )
    }

    private fun openWidgetsDialog() {
        showDialog(WidgetsDialog(this))
    }

    private fun openSettingsDialog() {
        showDialog(SettingsDialog(this))
    }

    private fun openForum() {
        val facebookIntent = Intent(Intent.ACTION_VIEW)
        facebookIntent.data = Config.FB_GROUP_URL.toUri()
        startActivity(facebookIntent)
    }

    private fun showMoreMenu() {
        morePopupWindow?.dismiss()

        val popupView = layoutInflater.inflate(R.layout.view_more_menu, binding.root, false)
        val popupWindow = PopupWindow(
            popupView,
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT,
            true
        ).apply {
            isOutsideTouchable = true
            elevation = 0f
            setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        }
        morePopupWindow = popupWindow

        popupView.findViewById<View>(R.id.more_widgets).setOnClickListener {
            popupWindow.dismiss()
            openWidgetsDialog()
        }
        popupView.findViewById<View>(R.id.more_settings).setOnClickListener {
            popupWindow.dismiss()
            openSettingsDialog()
        }
        popupView.findViewById<View>(R.id.more_forum).setOnClickListener {
            popupWindow.dismiss()
            openForum()
        }

        popupView.measure(
            View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED),
            View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
        )

        val menuLocation = IntArray(2)
        val anchorLocation = IntArray(2)
        val rootLocation = IntArray(2)
        binding.bottomNavigation.getLocationOnScreen(menuLocation)
        binding.navMore.getLocationOnScreen(anchorLocation)
        binding.root.getLocationOnScreen(rootLocation)

        val margin = resources.getDimensionPixelSize(R.dimen.more_menu_margin)
        val x: Int
        val y: Int
        if (isLandscape) {
            x = menuLocation[0] - rootLocation[0] + binding.bottomNavigation.width + margin / 2
            y = menuLocation[1] - rootLocation[1] + binding.bottomNavigation.height - popupView.measuredHeight - margin
        } else {
            x = anchorLocation[0] - rootLocation[0] + binding.navMore.width - popupView.measuredWidth
            y = anchorLocation[1] - rootLocation[1] - popupView.measuredHeight - margin
        }
        popupWindow.showAtLocation(binding.root, Gravity.NO_GRAVITY, x.coerceAtLeast(margin), y)
    }

    private fun openWidgetsDialogIfRequested(intent: Intent?) {
        val widgetId = intent?.getStringExtra(Config.EXTRA_WIDGET_ID) ?: return
        showDialog(WidgetsDialog(this, widgetId))
        intent.removeExtra(Config.EXTRA_WIDGET_ID)
    }

    companion object {
        private const val MY_PERMISSIONS_REQUEST_LOCATION = 1443
    }
}
